src/server/game/Anticheat/AnticheatMgr.h

```diff
@@ -0,0 +1,151 @@
+*
+* This program is distributed in the hope that it will be useful, but WITHOUT
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+* more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef SC_ACMGR_H
+#define SC_ACMGR_H
+
+#include "AnticheatData.h"
+#include "Common.h"
+#include "SharedDefines.h"
+#include "ScriptMgr.h"
+#include "Player.h"
+
+class ChatHandler;
+class AnticheatCheck;
+
+enum DetectionTypes
+{
+    SPEED_HACK_DETECTION            = 0x01,
+    FLY_HACK_DETECTION              = 0x02,
+    WALK_WATER_HACK_DETECTION       = 0x04,
+    JUMP_HACK_DETECTION             = 0x08,
+    TELEPORT_PLANE_HACK_DETECTION   = 0x10,
+    CLIMB_HACK_DETECTION            = 0x20
+};
+
+typedef std::unordered_map<ObjectGuid, AnticheatData> AnticheatPlayersDataMap;
+
+class AnticheatMgr
+{
+    friend class AnticheatCheck;
+
+    AnticheatMgr();
+    ~AnticheatMgr();
+
+    public:
+        static AnticheatMgr* instance()
+        {
+           static AnticheatMgr instance;
+           return &instance;
+        }
+
+        void StartHackDetection(Player* player, MovementInfo const& movementInfo, uint16 opcode);
+
+        void HandlePlayerLogin(Player* player, PreparedQueryResult result);
+        void HandlePlayerLogout(ObjectGuid const& guid);
+        void SavePlayerData(ObjectGuid const& guid, SQLTransaction& trans);
+
+        inline uint32 GetTotalReports(ObjectGuid const& guid) { return m_playerMap[guid].GetTotalReports(); }
+        inline float GetAverage(ObjectGuid const& guid) { return m_playerMap[guid].GetAverage(); }
+        inline uint32 GetTypeReports(ObjectGuid const& guid, ReportTypes type) { return m_playerMap[guid].GetTypeReports(type); }
+
+        // use sparingly, triggers a global anticheat save to DB, requires to lock player map and does some synchronous queries
+        bool AnticheatGlobalCommand(ChatHandler* handler);
+
+        void ResetDailyReportStates();
+
+        void LoadSettings();
+
+    private:
+        AnticheatData* GetPlayerData(ObjectGuid const& guid)
+        {
+            auto itr = m_playerMap.find(guid);
+            if (itr != m_playerMap.end())
+                return &itr->second;
+            return nullptr;
+        }
+
+        AnticheatPlayersDataMap m_playerMap;
+        std::vector<AnticheatCheck*> m_checks;
+
+        bool m_enabled;
+};
+
+#define sAnticheatMgr AnticheatMgr::instance()
+
+class AnticheatCheck
+{
+    public:
+        virtual ~AnticheatCheck() { }
+        virtual bool OnCheck(Player* player, MovementInfo const& movementInfo, uint16 opcode = 0) const = 0;
+
+        virtual void HackReport(Player* player) const = 0;
+
+    protected:
+        static AnticheatData* GetPlayerData(ObjectGuid const& guid) { return sAnticheatMgr->GetPlayerData(guid); }
```

Comment:
It's only an accessor, used in checks context, checks are only called on movementInfo handlers, which in turn only get called when player is logged in, data is already existing at this point.